### PR TITLE
Fix potential use-after-free and memory leak

### PR DESCRIPTION
**Pull request by xiaoyinl**
In function wait_for_async(), allocated async fds is freed if `SSL_get_all_async_fds`
fails, but later `fds` is used. Interestingly, it is not freed when everything succeeds.

### CODE REVIEW

**Comment by richsalz** [comment]
CLA on file, label removed; thanks.

File: apps/apps.c [hunk 1]
```diff
@@ -2633,6 +2634,7 @@ void wait_for_async(SSL *s)
         fds++;
     }
     select(width, (void *)&asyncfds, NULL, NULL, NULL);
+    OPENSSL_free(fds);
```
Comment (kroeckx) [inline apps/apps.c]: since fds has been increased, this won't actually work. You'll need to keep the original pointer.

Comment (xiaoyinl) [inline apps/apps.c]: @kroeckx Fixed. Thank you!

File: apps/apps.c [hunk 2]
```diff
@@ -2613,15 +2613,18 @@ void wait_for_async(SSL *s)
...
 
     if (!SSL_get_all_async_fds(s, NULL, &numfds))
         return;
     if (numfds == 0)
         return;
     fds = app_malloc(sizeof(OSSL_ASYNC_FD) * numfds, "allocate async fds");
+    origfds = fds;
     if (!SSL_get_all_async_fds(s, fds, &numfds)) {
         OPENSSL_free(fds);
+        return;
```
Comment (dot-asm) [inline apps/apps.c]: If `return` is right, then what does one need origfds for?

Comment (xiaoyinl) [inline apps/apps.c]: So do you want me to move `origfds = fds` below this line?

File: apps/apps.c [hunk 3]
```diff
@@ -2633,6 +2636,7 @@ void wait_for_async(SSL *s)
         fds++;
     }
     select(width, (void *)&asyncfds, NULL, NULL, NULL);
+    OPENSSL_free(origfds);
```
Comment (bernd-edlinger) [inline apps/apps.c]: If the above loop avoids pointer arithmetic, then the origfds would not be needed?
right?  For instance like this:
```
while (numfds > 0) {
  numfds--
  if (width <= (int)fds[numfds])
     width = (int)fds[numfds] + 1;
  openssl_fdset((int)fds[numfds], &asyncfds);
}
select(width, (void *)&asyncfds, NULL, NULL, NULL);
OPENSSL_free(fds);
```

Comment (paulidale) [inline apps/apps.c]: I thought of this change too.  I think it's easier to make sense of but I wasn't sure it was worth the effort.  Given a second notice, I say go for it.

Comment (xiaoyinl) [inline apps/apps.c]: Done. Thank you for the suggestion!

**Comment by dot-asm** [comment]
Since there were new commits and I've added 1.1.0 label, I consider original approval void, and for this reason add "pending 2nd review" instead of "ready".

**Comment by richsalz** [comment]
Feel free to merge. @paulidale

**Comment by paulidale** [comment]
Should we wait for @bernd-edlinger to reapprove as well?

I'm happy to do the merge.

**Comment by paulidale** [comment]
Merged.

**Comment by xiaoyinl** [comment]
Thanks!

**Comment by InfoHunter** [comment]
One disadvantage of 'squash' commits is that the original author's information will be lost, this would be hard to trace changes by using something like 'git blame' in the future...

**Comment by xiaoyinl** [comment]
I agree with @InfoHunter.
@paulidale @levitte: Next time can I squash the commits and do a force push, and you merge that single commit? In this way, the author's info is kept and the history is clean.

**Comment by dot-asm** [comment]
> One disadvantage of 'squash' commits is that the original author's information will be lost,

I recall very clearly squashed #3242 and it didn't loose original author. [And I think I did more squashing without loosing it.] I don't recall that I actually did anything special to achieve that. Problem with contributor squashing is that *formally* speaking reviewer *has to* restart review effectively from scratch. Because there is no way to be *sure* that it was just squash. So that formally speaking correct course of action is to always let reviewer (or more accurately committer) to do the squashing. And from my experience it doesn't [have to] loose the original author. In other words I'd rather see that contributor information is preserved by committer's squash than contributor's squash...

**Comment by richsalz** [comment]
The squash script we use can lose authorship if not careful, but that's a script bug.  (My script, my bug, and I've done it a couple of times to people, sorry.)  Often a PR will go through multiple iterations and fixes and we don't really want all those "you have a leak here" fixups to be in the main commit log.  Yes, it would be good if authors did the "right" squashing before we merge it.  Right has no hard and fast rule. :)  Like much of software development, it's about taste and style more than strict engineering.

TL;DR  Shrug :)

**Comment by InfoHunter** [comment]
It would be perfect if the bug in squash script is going to be fixed : )

**Comment by dot-asm** [comment]
> it would be good if authors did the "right" squashing before we merge it.

No, it would be *convenient* [for reviewer]. But convenience is not the purpose of review process, quality ensurance is. At the same time it's desired that it's done with least required effort. So that for big/complex enough request it *might* be plain inappropriate to let contributor squash. [Do note "might", as I'm not saying that there might be situations when contrary would be appropriate.]

> Right has no hard and fast rule.

Couldn't agree more, but there *are* [quite natural] priorities.

> it's about taste and style more than strict engineering.

But when and how to squash has nothing to do with software engineering. Coding can be about taste and style, but maintaining *shared* repository is better be more strict than coding style guide is. Do note that I'm not saying that it has to be rigid, only that you can't really draw sign of equivalence between them, nor that it can be *as* permissive.

**Comment by kaduk** [comment]
> No, it would be convenient [for reviewer]. But convenience is not the purpose of review process, quality ensurance is. At the same time it's desired that it's done with least required effort. So that for big/complex enough request it might be plain inappropriate to let contributor squash. [Do note "might", as I'm not saying that there might be situations when contrary would be appropriate.]

It's not clear that the reviewer needs to restart the review from scratch if the VCS tooling can help.  That is, 'git fetch github refs/pull/NNNN/head' can get a local copy of the code originally reviewed, which can have a branch name assigned to is, so that when the contributor's squash is available, the new code can be fetched and 'git diff' used to detect any changes.

**Comment by dot-asm** [comment]
Maybe it would be appropriate to move to #3963? At least it's more related and is not closed. I'll add a note including this remark.

**Comment by richsalz** [comment]
I believe I addressed the ghmerge bug.  When doing a squash, the approach is now to do a "rebase -i"
@paulidale take a look.

**Comment by paulidale** [comment]
The updated script looks okay.

**Pull request closed without merging**